### PR TITLE
Add Support for chatglm2

### PR DESCRIPTION
**Pull request by InkdyeHuang**
1.edit rotary_embedding to support chatglm2 and use vllm's multi-query attention for chatglm2
2. I have test the result
a. Inference result alignment.
b .Performance result: greedy mode when input_token_name=250, max_output_token=50, greedy mod; torch Throughput 22samples/s ; vllm Throughput 37.4samples/s

### CODE REVIEW

**Comment by InkdyeHuang** [comment]
1. Inference result alignment.
2 .Performance result: when input_token_name=250, max_output_token=50; torch Throughput 22samples/s ; vllm Throughput 37.4samples/s

File: csrc/pos_encoding_kernels.cu
Comment (magiclogy-) [inline csrc/pos_encoding_kernels.cu:1]: I tested your code, but it failed to produce meaningful words.

After hours of debugging, I found a bug in the rotary embedding for keys:

```
template<typename scalar_t>
__global__ void rotary_embedding_neox_kernel(
...
  const int nk = num_kv_heads * embed_dim;
  for (int i = threadIdx.x; i < nk; i += blockDim.x) {
...
    const scalar_t k_x = key[token_head + x_index];
    const scalar_t k_y = key[token_head + y_index];
    key[out_x] = k_x * cos - k_y * sin;
    key[out_y] = k_y * cos + k_x * sin;
  }
}
```

In line 61-62, `k_x` should be `key[out_x]` and `k_y` should be `key[out_x]`.

**Comment by GoHomeToMacDonal** [comment]
> > 1.edit rotary_embedding to support chatglm2 and use vllm's multi-query attention for chatglm2 2. I have test the result a. Inference result alignment. b .Performance result: greedy mode when input_token_name=250, max_output_token=50, greedy mod; torch Throughput 22samples/s ; vllm Throughput 37.4samples/s
> 
> Found (prompt)context decoding process aligned but tokens'generation not, possibly due to the cuda kernel.

The key problem is the cos_sin_cache is not aligned with chatglm2, because chatglm2 use mixed-precision to compute it. I will provide the corresponding code later. Maybe you can load it from chatglm2.

File: vllm/model_executor/layers/attention.py
Comment (GoHomeToMacDonal) [inline vllm/model_executor/layers/attention.py:1]: Please note that ChatGLM2 adopts mixed-precision floats to compute the cos_sin_cache for rotary embedding, but the original vLLM does not. 

The implementation of ChatGLM2 is 
```
inv_freq = 1.0 / (base**(torch.arange(0, rotary_dim, 2) / rotary_dim)).to(dtype=torch.float16)
t = torch.arange(max_position).float().to(dtype=torch.float16)
freqs = torch.einsum("i,j -> ij", t, inv_freq).float()
```

The implementation of vLLM is 
```
inv_freq = 1.0 / (base**(torch.arange(0, rotary_dim, 2) / rotary_dim))
t = torch.arange(max_position).float()
freqs = torch.einsum("i,j -> ij", t, inv_freq.float())
```



This difference will make the output tokens slightly different when the generated text is long enough.

**Comment by zhuohan123** [comment]
Again, thank you for your contribution! We merged https://github.com/vllm-project/vllm/pull/1261 and the current main branch supports ChatGLM now. Let us know if the main branch does not look good and feel free to propose any changes!

**Pull request closed without merging**